services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crossregion/S3CrossRegionSyncClient.java

```diff
@@ -15,80 +15,110 @@
 
+import java.util.Map;
 import java.util.Optional;
-import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.function.Function;
+import java.util.function.Supplier;
 import software.amazon.awssdk.annotations.SdkInternalApi;
 import software.amazon.awssdk.awscore.AwsRequestOverrideConfiguration;
-import software.amazon.awssdk.endpoints.Endpoint;
+import software.amazon.awssdk.regions.Region;
 import software.amazon.awssdk.services.s3.DelegatingS3Client;
 import software.amazon.awssdk.services.s3.S3Client;
-import software.amazon.awssdk.services.s3.endpoints.S3EndpointParams;
 import software.amazon.awssdk.services.s3.endpoints.S3EndpointProvider;
+import software.amazon.awssdk.services.s3.internal.crossregion.endpointprovider.BucketEndpointProvider;
+import software.amazon.awssdk.services.s3.model.HeadBucketRequest;
+import software.amazon.awssdk.services.s3.model.S3Exception;
 import software.amazon.awssdk.services.s3.model.S3Request;
 
+/**
+ * Decorator S3 Sync client that will fetch the region name whenever there is Redirect 301 error due to cross region bucket
+ * access.
+ */
 @SdkInternalApi
 public final class S3CrossRegionSyncClient extends DelegatingS3Client {
+
+    private static final String AMZ_BUCKET_REGION_HEADER = "x-amz-bucket-region";
+    public static final int REDIRECT_STATUS_CODE = 301;
+    private final Map<String, Region> bucketToRegionCache = new ConcurrentHashMap<>();
+
     public S3CrossRegionSyncClient(S3Client s3Client) {
         super(s3Client);
     }
 
+    private static <T extends S3Request> Optional<String> bucketNameFromRequest(T request) {
+        return request.getValueForField("Bucket", String.class);
+    }
+
     @Override
     protected <T extends S3Request, ReturnT> ReturnT invokeOperation(T request, Function<T, ReturnT> operation) {
 
-        Optional<String> bucket = request.getValueForField("Bucket", String.class);
-
-        if (bucket.isPresent()) {
-            try {
-                return operation.apply(requestWithDecoratedEndpointProvider(request, bucket.get()));
-            } catch (Exception e) {
-                handleOperationFailure(e, bucket.get());
+        Optional<String> bucketRequest = bucketNameFromRequest(request);
+        if (!bucketRequest.isPresent()) {
+            return operation.apply(request);
+        }
+        String bucketName = bucketRequest.get();
+        try {
+            if (bucketToRegionCache.containsKey(bucketName)) {
+                return operation.apply(requestWithDecoratedEndpointProvider(request, regionSupplier(bucketName)));
+            }
+            return operation.apply(request);
+        } catch (S3Exception exception) {
+            if (exception.statusCode() == REDIRECT_STATUS_CODE) {
+                updateCacheFromRedirectException(exception, bucketName);
+                return operation.apply(requestWithDecoratedEndpointProvider(request, regionSupplier(bucketName)));
```

Comment:
If `fetchBucketRegion` is called because the redirect exception does not contain the region, it seems we will not update the cache?